src/maxtree.jl

```diff
@@ -0,0 +1,881 @@
+        getindex(LinearIndices(image), center)
+
+    # initialization of the image parent
+    parents = fill!(maxtree.parents, 0) |> vec
+    roots = fill!(similar(parents), 0) # temporary array containing current roots
+    rootpath = Vector{Int}() # temp array to hold the path to the current root within ancestors
+    cindexes = CartesianIndices(maxtree.parents)
+
+    # traverse the array in reversed order (from highest value to lowest value)
+    @inbounds for p in Iterators.Reverse(maxtree.traverse)
+        p_ci = cindexes[p]
+        parents[p] = roots[p] = p # it's a new root
+
+        for i in eachindex(neighbors)
+            # check that p is in the mask and that it's neighbor is a valid image pixel
+            (#=mask[p] && =#isvalid_offset(neighbors[i], p_ci, maxtree.axes)) || continue
+            index = p + neighbor_offsets[i] # linear index of the neighbor
+            (parents[index] == 0) && continue # ignore neighbor without parent (= its value is lower)
+            root = rootpath!(rootpath, roots, index) # neighbor's root
+            if (root != p) || (length(rootpath) > 1) # attach neighbor's root to the p
+                parents[root] = p
+                roots[rootpath] .= p # also attach the ancestral nodes of neighbor to p
+            end
+        end
+    end
+    return canonize!(maxtree, image)
+end
+
+# generates a vector of offsets to the CartesianIndex that defines
+# the neighborhood of the pixel in an N-dimensional image.
+# `connectivity` is the maximal number of orthogonal steps
+# (+1/-1 to a single dimension) that are required to reach a neighbor.
+function neighbor_cartesian_offsets(::Type{CartesianIndex{N}}, connectivity::Integer) where N
+    cis = CartesianIndices(ntuple(_ -> 3, Val{N}()))
+    offsets = [Tuple(ci) .- 2 for ci in cis]
+    return filter!(offset -> 0 < sum(abs, offset) <= connectivity, vec(offsets))
+end
+
+"""
+    MaxTree(image::AbstractArray; [connectivity=1], [rev=false]) -> MaxTree
+
+Constructs the *max-tree* of the `image`.
+
+# Arguments
+- `connectivity::Integer=1`: defines the pixel neighborhood used to construct
+  the connected components. The value is the maximum number of orthogonal steps
+  to reach a neighbor. In 2D, it is 1 for a 4-neighborhood and 2 for a
+  8-neighborhood. See [`rebuild!](@ref).
+- `rev::Bool=false`: if `false`, the max-tree is traversed from the darkest
+  (the root node) to the brightest, otherwise it's traversed from the brightest
+  (the root) to the darkest.
+
+# Examples
+We create a small sample image (Figure 1 from [4]) and build the max-tree.
+
+```jldoctest
+julia> image = [15 13 16; 12 12 10; 16 12 14]
+3×3 Array{Int64,2}:
+ 15  13  16
+ 12  12  10
+ 16  12  14
+
+julia> mtree = MaxTree(image, connectivity=2)
+MaxTree{2,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}}}((Base.OneTo(3), Base.OneTo(3)), false, [4 2 4; 8 2 8; 2 2 2], [8, 2, 5, 6, 4, 9, 1, 3, 7])
+```
+"""
+function MaxTree(image::AbstractArray{<:Number, N};
```

Comment:
should also dispatch on `GenericGrayImage`

Reply:
I'm trying to test `GenericGrayImage`, but it fails with
```
 Expression: MaxTree(Gray.(A ./ 255)) isa MaxTree{2}
  MethodError: no method matching isless(::Gray{Float64}, ::Gray{Float64})
  Closest candidates are:
    isless(!Matched::Missing, ::Any) at missing.jl:87
    isless(::Any, !Matched::Missing) at missing.jl:88
  Stacktrace:
   [1] lt(::Base.Order.ForwardOrdering, ::Gray{Float64}, ::Gray{Float64}) at ./ordering.jl:49
   [2] sort!(::Array{Int64,1}, ::Base.Sort.QuickSortAlg, ::Base.Order.Perm{Base.Order.ForwardOrdering,Array{Gray{Float64},1}}) at ./ordering.jl:57
   [3] #sortperm!#12(::Base.Sort.QuickSortAlg, ::Function, ::Function, ::Bool, ::Base.Order.ForwardOrdering, ::Bool, ::typeof(sortperm!), ::Array{Int64,1}, ::Array{Gray{Float64},1}) at ./sort.jl:907
   [4] #sortperm! at ./none:0 [inlined]
   [5] rebuild!(::MaxTree{2}, ::Array{Gray{Float64},2}, ::Array{CartesianIndex{2},1}) at /home/ge68wan/.julia/dev/ImageMorphology/src/maxtree.jl:168
   [6] #MaxTree#35(::Int64, ::Bool, ::Type{MaxTree}, ::Array{Gray{Float64},2}) at /home/ge68wan/.julia/dev/ImageMorphology/src/maxtree.jl:239
   [7] MaxTree(::Array{Gray{Float64},2}) at /home/ge68wan/.julia/dev/ImageMorphology/src/maxtree.jl:227
```

Shouldn't there be `isless()` defined forr grayscale?